BIOIN-1558 srsnv cv by chromosome

orihir opened the pull request:
## Overview
This PR adds support for training SRSNV model with a cross-validation (CV) scheme. CV is needed when SRSNV inference is used on the same featuremaps used for training, because without CV there's a danger of overfitting. 

## How to use
To use CV, need to set the new parameter `num_CV_folds` to the number of folds required. By default, this is 1, i.e., no CV (regular train-test split). Each read in the training featuremap dataframe is assigned a fold, stored in the `fold_id` column. Model `k` associated with fold `k` is trained on all data where `fold_id != k`, and validated on `fold_id == k`. By default, folds are split by chromosome, to make sure there is no leakage between training and validation sets (`fold_id`'s can be assigned randomly by setting `split_folds_by` to `random` rather than `chrom`. With random fold-splitting, at inference time an ensemble of all `num_CV_folds` models is used for inference). 

`fold_id` can have one of 3 types of values.
- A number `k` in `0, 1, ..., num_CV_folds-1`. These reads are in the training set of all models except model `k`. 
- -1. These reads are in the training set when `num_CV_folds==1`, i.e., when using regular train-test split. The test set reads in this case have `fold_id=0`. 
- `np.nan`. These reads are not included in any training set, and so (an ensemble of) all models can be used for inference. 

The new function `k_fold_predict_proba` can be used to evaluate the trained models on a dataset (it accepts a list of models, a dataset, and the name of the column containing fold_ids, and evaluates the models according to these rules. 

## Other changes in this PR
- Categorical features are defined together with all possible values, ensuring that they are consistently encoded by the model during training, validation, and inference. 
- Models are saved in a joblib file that contains relevant meta-data, like the parameters used in training (which include the mapping of chromosomes to folds, and the details of all categorical variables), and also the interpolating function that maps ML_qual to SNVQ. 
- Hyperparameters were changes to include more training rounds and use early-stopping to reduce overfitting.
- Small bug fixes, including in the calculation of the residual snv rate.

orihir committed:
add cross validation support to srsnv inference

Also, add support of new categorical context features (next_i, prev_o) for NYGC data
which has old categorical context features (trinuc_context_with_alt).

orihir committed:
Fix header bug when prefiltering on mixed reads

orihir committed:
Fix category order in features

orihir committed:
srsnv training cross validation support

kenissur committed:
Changed flag to split_folds_randomly

orihir committed:
Test precommit

orihir committed:
fix formatting

orihir committed:
fix formatting of srsnv_inference

kenissur committed:
bugfix

kenissur committed:
rename data_df to featuremap_df

kenissur committed:
Added qual column to dataframe instead of separate file

orihir committed:
srsnv hyperparameters add early stopping

orihir committed:
customize categorical feature values

Use a dict to define categorical features together with the possible category values.
each category is defined by a dict:
{cat_name: [[cat_val1, cat_val2, ...], ordered}
where
- cat_name [str]: category name
- [cat_val1, cat_val2, ...] [list]:  category values
- ordered [bool]: indicates whether categories are ordered or not (when in doubt, use False)

orihir committed:
pass srsnv dataset params via json file

orihir committed:
Save model with metadata

Changes in this commit:
- Model(s) are saved as a single joblib file, which includes also params and quality_interpolation_function.
- Saving model(s) as json file is made optional (by adding argument --save_model_jsons).
- Chromosome sizes are read from reference.dict file, rather than hard-coded.
- Bug fix in calculation of downsampling rate for LoD calculation of test set when using cross-validation
- Number of folds is added to saved params.

orihir committed:
CV in inference conforms to training

orihir committed:
SRSNV inference supports random fold splitting

orihir committed:
clean calculation of residual snv rate

This commit cleans the way that the effective_bases_covered and
residual_snv_rate_no_filter are calculated in the srsnv report / plotting utils. Before,
residual_snv_rate_no_filter was calculated as the ratio of  the FP train/test dataset size
in the numerator of and effective_bases_covered as the denominator. The latter
included a factor of the downsampling rate (from the number of entries in the FP
featuremap file to the [smaller] dataset size). After this commit, the numerator is
the number of reads in the fp featuremap directly, and the downsampling rate is
removed from the denominator.

This commit also fixes a factor of 2 in the snv_rate that was caused by calculating the
downsampling rate as the ratio : (TP + FP dataset set) / (FP entry number).

orihir committed:
Restrict training to chromosomes in TP bed file

This is needed for splitting the data into folds for cross-validation:
fold-splitting is done so that folds will all have approximately the
same size. If training is done only on a few chromosomes, should make
sure that they are not in the same fold, and that the fold sizes
are approximately equal.

kenissur committed:
Updated MRD analysis notebook to modified SRSNV output (featuremap_df_file)

kenissur committed:
Fixed handling of nan values in plots in MRD notebook

kenissur committed:
removed hard coded formatting of chromosome names "chr{x}", changed handling of reference_dict and tp_regions_bed_file

kenissur committed:
fixed (?) quality interpolation

orihir committed:
srsnv correct qual calculation

orihir committed:
Adjust srsnv tests

kenissur commented on ugvc/mrd/srsnv_inference_utils.py:
```diff
@@ -39,48 +45,65 @@ def _vcf_getter(variant, field):
         pre_filter: str = None,
-        test_set_mrd_simulation_dataframe_file: str = None,
         low_qual_threshold: float = LOW_QUAL_THRESHOLD,
+        # new_categorical_features: list[str] = None, # <- HERE, make sure these are not used!
```
Remove the comment please

kenissur commented on ugvc/mrd/srsnv_inference_utils.py:
```diff
@@ -164,22 +183,48 @@ def process_records(self, records: list[pysam.VariantRecord]) -> list[pysam.Vari
+        #             predicted_probability[df["fold"].isna().index, :] += (
+        #                 model.predict_proba(df.loc[df["fold"].isna(), features_for_model]) / self.num_folds
+        #             )  # chromosomes not in any fold, predict with all models and take mean
+
```
Please remove comments

kenissur commented on ugvc/mrd/srsnv_inference_utils.py:
```diff
@@ -164,22 +183,48 @@ def process_records(self, records: list[pysam.VariantRecord]) -> list[pysam.Vari
-        predicted_probability = self.model.predict_proba(df)
+        features_for_model = self.numerical_features + self.categorical_features_names
+        if self.num_folds == 1:
+            df["fold_id"] = 0
```
A bit petty but maybe use a constant FOLD_ID?

kenissur commented on ugvc/mrd/srsnv_inference_utils.py:
```diff
@@ -220,54 +265,53 @@ def extract_categories_from_parquet(parquet_path):
     low_qual_threshold: float = LOW_QUAL_THRESHOLD,
     process_number: int = 0,
+    # num_folds: int = 0,
+    # chrom_folds_path: str = None,
```
Comments

kenissur committed:
bugfix (ML_qual_1_test -> ML_qual_1_test)

kenissur committed:
fixed FILTERED_COUNT normalization

orihir committed:
srsnv small bug fixes

1. Fix bug in create_report when output path contains the string 'test'
2. Read model_parameters from file if provided
3. define 'is_mixed' column in create_report and then use it in simple LoD filters
4. Fix typo in logging of normalization constants

orihir committed:
srsnv tests updated to work with cross-validation

kenissur committed:
Final code cleanup
